src/couch_prometheus/src/couch_prometheus_server.erl

```diff
@@ -221,36 +223,155 @@ get_message_queue_stats() ->
+    Count = length(Sorted),
+    Quantiles = [
+        {[{quantile, <<"0.5">>}], lists:nth(round(Count * 0.5), Sorted)},
+        {[{quantile, <<"0.9">>}], lists:nth(round(Count * 0.9), Sorted)},
+        {[{quantile, <<"0.99">>}], lists:nth(round(Count * 0.99), Sorted)}
+    ],
+    SumStat = to_prom(Metric ++ ["_sum"], lists:sum(Sorted)),
+    CountStat = to_prom(Metric ++ ["_count"], length(Sorted)),
+    MinStat = to_prom(Metric ++ ["_min"], hd(Sorted)),
+    MaxStat = to_prom(Metric ++ ["_max"], lists:nth(Count, Sorted)),
+    to_prom(Metric, summary, Desc, Quantiles) ++ [SumStat, CountStat, MinStat, MaxStat].
 
 get_run_queue_stats() ->
     %% Workaround for https://bugs.erlang.org/browse/ERL-1355
-    {Normal, Dirty} =
-        case erlang:system_info(dirty_cpu_schedulers) > 0 of
-            false ->
-                {statistics(run_queue), 0};
-            true ->
-                [DCQ | SQs] = lists:reverse(statistics(run_queue_lengths)),
-                {lists:sum(SQs), DCQ}
-        end,
+    {SQ, DCQ} = chttpd_node:run_queues(),
     [
-        to_prom(erlang_scheduler_queues, gauge, "the total size of all normal run queues", Normal),
+        to_prom(erlang_scheduler_queues, gauge, "the total size of all normal run queues", SQ),
         to_prom(
             erlang_dirty_cpu_scheduler_queues,
             gauge,
             "the total size of all dirty CPU scheduler run queues",
-            Dirty
+            DCQ
         )
     ].
 
+get_distribution_stats() ->
+    % each distribution metric has a different type,
+    % so expose each as a different metric with the erlang
+    % node as a label.
+    % This is the inverse of the structure returned by
+    % inet:getstat/1.
+
+    % this is a bit hairy. This fold accumulates
+    % a dictionary keyed on the socker stat_option (https://www.erlang.org/doc/man/inet.html#getstat-2)
+    % where the value is a list of labels/value pairs for that stat
+    % e.g.
+    % recv_oct: [{[{node="node2@127.0.0.1"}], 30609}, {[{node="node3@127.0.0.1"}], 28392}]
+    % recv_cnt: [{[{node="node2@127.0.0.1"}], 123}, {[{node="node3@127.0.0.1"}], 134}]
+    DistStats = lists:foldl(
+        fun({Node, Socket}, Acc) ->
+            try inet:getstat(Socket) of
+                {ok, Stats} ->
+                    % For each Key/Value pair in Stats, append
+                    % an entry for the current Node to the result.
+                    % This relies on lists:foldl returning the final
+                    % accumulated value
+                    lists:foldl(
+                        fun({StatOption, Value}, Acc0) ->
+                            dict:append(StatOption, {[{node, Node}], Value}, Acc0)
+                        end,
+                        Acc,
+                        Stats
+                    )
+            catch
+                _:_ ->
+                    % no result so just continue
+                    Acc
+            end
+        end,
```

Comment:
this is probably the gnarliest bit of the PR - suggestions for a cleaner implementation welcome!

Reply:
thanks - I've pulled out a `get_sock_stats` function which uses a map instead of s dict as you suggested, which I think is cleaner.